projects/go-ethereum/Dockerfile

```diff
@@ -16,28 +16,10 @@
 
 FROM gcr.io/oss-fuzz-base/base-builder
 
-RUN go get github.com/ethereum/go-ethereum
-RUN go get golang.org/x/sys/cpu
-RUN go get github.com/go-stack/stack
-RUN go get github.com/deckarep/golang-set
-RUN go get github.com/golang/snappy
-RUN go get github.com/gorilla/websocket
-RUN go get github.com/hashicorp/golang-lru
-RUN go get github.com/holiman/uint256
-RUN go get github.com/olekukonko/tablewriter
-RUN go get github.com/prometheus/tsdb/fileutil
-RUN go get github.com/shirou/gopsutil/cpu
-RUN go get github.com/steakknife/bloomfilter
-RUN go get github.com/syndtr/goleveldb/leveldb
-RUN go get github.com/syndtr/goleveldb/leveldb/errors
-RUN go get github.com/syndtr/goleveldb/leveldb/filter
-RUN go get github.com/syndtr/goleveldb/leveldb/opt
-RUN go get github.com/syndtr/goleveldb/leveldb/util
-RUN go get github.com/VictoriaMetrics/fastcache
-RUN go get github.com/aristanetworks/goarista/monotime
-RUN go get github.com/pborman/uuid
-RUN go get github.com/rjeczalik/notify
-RUN go get github.com/google/gofuzz
+RUN mkdir -p /root/go/src/github.com/ethereum
+RUN (cd /root/go/src/github.com/ethereum/ && \
+	git clone --single-branch --depth=1 https://github.com/ethereum/go-ethereum )
+RUN (cd /root/go/src/github.com/ethereum/go-ethereum/ && go build ./... )
```

Comment:
This looks wrong, we don't build inside Dockerfile, we do the build in build.sh file. (as we want build.sh to use any set CFLAGS/CXXFLAGS, etc).
Also, why do we need to clone inside /root/go/src/github.com/ethereum, we should just change build/oss-fuzz.sh to account for source paths,

WORKDIR can be set to $SRC/go-ethereum.

Reply:
I agree it looks a bit odd. The idea here isn't to build, per se (beacuse we're totally going to ignore the built binaries). 

> Also, why do we need to clone inside /root/go/src/github.com/ethereum

This is an alternative to `go get github.com/ethereum/go-ethereum`. Both of these places the `go-ethereum` code into `$GOPATH` (`~/go`), but

1. The `go get` syntax uses certain rules to determine which version to get, and would normally get the latest stable version (but we prefer the fuzzer to run on latest head)
2. The `go get` syntax fetches a lot more git history, which we can skip by `git clone`. 

After we have `go-ethereum` in `$GOPATH`, the `go build ./...` is used to replace the explicit `go get ...` in the original Dockerfile. It triggers the the modules to be resolved and downloaded. 

The actual building of fuzzer targets happen in `build.sh`, as before. It would be possible to do more of this work inside `build.sh` and not "pre-load" as much in the Dockerfile. 

I did notice, however, that whereas the go-lang builders properly resolved all modules, the `go-fuzz` step was not as clever about it, and spat out some errors - on exactly those repositories that was originally explcitly added in the Dockerfile. It seems that `go-fuzz` expects the deps to be in `$GOPATH`. 

I can look into alternative ways of doing this, but I'm curios about a more high-level question: 

1. When are the docker images rebuilt? What is the expected lifecycle of a fuzzer docker image? 
2. And respectively, when is the `build.sh` rerun, and what is the lifecycle of that? 

Depending on the answer to that -- but I suspect the 'proper' way to do it would be to not "preload" anything at all inside the docker image, aside from `go` and `git`, and fetch all source+deps in the build step?

Reply:
Also, maybe related. When I try to access the coverage report, it fails. Could the root cause of that be the code is located in `/root/go` instead of `$SRC/` ?

Reply:
If we need to make it work with tools that aren't Go module-aware, we can use `go mod vendor` to populate the vendor/ directory with all dependencies. This will also download the dependencies AFAIK.

Reply:
Actually, the docker image doesn't put the base (go-ethereum) into $GOPATH, but instead just dumps it into `$SRC`, then it seems like all tools work well. I'm going to update this PR to be even simpler, once I've ran all the local tests